lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -22,92 +56,98 @@ def get_params():
-
-    tar = tarfile.open(tar_file)
-    # There should be only one member inside this tar file i.e. null
-    for member in tar.getmembers():
-        assert member.name == "null"
-
-    tar_file.unlink()
+
+    def test_create_tar(self, agent_setup, mock_tar):
+        """Test create tar file"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
+            redis_server=None,
+            logger=None,
+        )
+        tmp_dir = agent_setup["tmp"]
+        tar_file = "test.tar.xz"
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 0
+        assert cp.stdout == b""
+
+    def test_create_tar_ignore_warnings(self, agent_setup, mock_tar_no_warnings):
+        """Test if we can suppress the errors raised during the tar creation"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
+            redis_server=None,
+            logger=logger,
+        )
+        tmp_dir = agent_setup["tmp"]
+        tar_file = "test.tar.xz"
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 1
+        assert cp.stdout == b"Some error running tar"
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file), retry=True)
+        assert cp.returncode == 0
+        assert cp.stdout == b""
+
+    def test_create_tar_failure(self, agent_setup, mock_tar_failure):
+        """Test if we can suppress the errors raised during the tar creation"""
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
+            redis_server=None,
+            logger=logger,
+        )
+        tmp_dir = agent_setup["tmp"]
+        tar_file = "test.tar.xz"
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 1
+        assert cp.stdout == b"Some error running tar"
+
+
+class TestSendDirectory:
+    """Test create_tar in send directory of tool meister"""
+
+    def test_tar_create_success(self, agent_setup, mock_tar, caplog):
+        """This test should pass the tar creation in send directory"""
+
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
+            redis_server=None,
+            logger=logger,
+        )
+        directory = agent_setup["tmp"] / f"{TestCreateTar.tm_params['hostname']}"
+        failures = tm._send_directory(directory, "", "")
+        assert "Failed to create an empty tar" not in caplog.text
+        assert failures == 1
```

Comment:
Why are we expecting a failure?  This should have a big ole code comment explaining it...either that, or you should be mocking `md5sum()`, `shutil.rmtree()`, `Path.unlink()`, etc., as well as using the `Requests` mocking for `PUT`, to drive this case to an actual successful return.

(I realize that testing `_send_directory()` is bear, but this test is nominally a "success case", so we need to make it clear that it is not.)

Also, for this test and the next one, you should be mocking out `_create_tar()` -- that is, the previous unit tests should be sufficient to test that function's code -- what we want to be testing here is the code which _calls_ the new function.  The mock should be asserting that the arguments it receives are appropriate, and it should generate the appropriate return value directly, without relying on mocking `subprocess.run()` to do it.  (I.e., the "unit" here should be `_send_directory()`, not "`_send_directory()` plus everything it calls".)

Reply:
Yeah I had to add quite a lot of mocks for `send_directory` to succeed.